drivers/serial/uart_rpi_pico.c

```diff
@@ -14,62 +15,313 @@
+	uart_hw->imsc &= ~(UART_UARTIMSC_OEIM_BITS |
+					   UART_UARTIMSC_BEIM_BITS |
+					   UART_UARTIMSC_PEIM_BITS |
+					   UART_UARTIMSC_FEIM_BITS |
+					   UART_UARTIMSC_RTIM_BITS);
+}
+
+static int uart_rpi_irq_is_pending(const struct device *dev)
+{
+	return !!(uart_rpi_irq_rx_ready(dev) || uart_rpi_irq_tx_ready(dev));
+}
+
+static int uart_rpi_irq_update(const struct device *dev)
+{
+	return 1;
+}
+
+static void uart_rpi_irq_callback_set(const struct device *dev,
+					uart_irq_callback_user_data_t cb,
+					void *cb_data)
+{
+	struct uart_rpi_data * const data = DEV_DATA(dev);
+
+	data->irq_cb = cb;
+	data->irq_cb_data = cb_data;
+}
+
+static void uart_rpi_isr(const struct device *dev)
+{
+	struct uart_rpi_data * const data = DEV_DATA(dev);
+
+	if (data->irq_cb) {
+		data->irq_cb(dev, data->irq_cb_data);
+	}
+}
+
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
 static const struct uart_driver_api uart_rpi_driver_api = {
 	.poll_in = uart_rpi_poll_in,
 	.poll_out = uart_rpi_poll_out,
+	.err_check = uart_rpi_err_check,
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill = uart_rpi_fifo_fill,
+	.fifo_read = uart_rpi_fifo_read,
+	.irq_tx_enable = uart_rpi_irq_tx_enable,
+	.irq_tx_disable = uart_rpi_irq_tx_disable,
+	.irq_tx_ready = uart_rpi_irq_tx_ready,
+	.irq_rx_enable = uart_rpi_irq_rx_enable,
+	.irq_rx_disable = uart_rpi_irq_rx_disable,
+	.irq_tx_complete = uart_rpi_irq_tx_complete,
+	.irq_rx_ready = uart_rpi_irq_rx_ready,
+	.irq_err_enable = uart_rpi_irq_err_enable,
+	.irq_err_disable = uart_rpi_irq_err_disable,
+	.irq_is_pending = uart_rpi_irq_is_pending,
+	.irq_update = uart_rpi_irq_update,
+	.irq_callback_set = uart_rpi_irq_callback_set,
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
 };
 
-#define RPI_UART_INIT(idx)						\
-	PINCTRL_DT_INST_DEFINE(idx);					\
-	static const struct uart_rpi_config uart_rpi_cfg_##idx = {	\
-		.uart_dev = (uart_inst_t *)DT_INST_REG_ADDR(idx),	\
-		.baudrate = DT_INST_PROP(idx, current_speed),		\
-		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),		\
-	};								\
-									\
-	DEVICE_DT_INST_DEFINE(idx, &uart_rpi_init,			\
-				NULL,					\
-				NULL,					\
-				&uart_rpi_cfg_##idx, PRE_KERNEL_1,	\
-				CONFIG_SERIAL_INIT_PRIORITY,		\
-				&uart_rpi_driver_api);
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+
+#define RPI_UART_CREATE_CFG(idx)	\
+	static const struct uart_rpi_config uart##idx##_rpi_config = {	\
+		.uart_dev = uart##idx,	\
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),	\
+		.irq_config_func = uart##idx##_rpi_irq_config_func	\
+	}
+
+#define RPI_UART_CONFIG_FUNC(idx)	\
+	static void uart##idx##_rpi_irq_config_func(const struct device *port)	\
+	{	\
+		IRQ_CONNECT(DT_INST_IRQN(idx),	\
+			    DT_INST_IRQ(idx, priority),	\
+			    uart_rpi_isr,	\
+			    DEVICE_DT_INST_GET(idx), 0);	\
+		irq_enable(DT_INST_IRQN(idx));	\
+	}
+
+#else
+
+#define RPI_UART_CREATE_CFG(idx)	\
+	static const struct uart_rpi_config uart##idx##_rpi_config = {	\
+		.uart_dev = uart##idx,	\
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),	\
```

Comment:
this is duplicating initializer fields unnecessarily, better use something like:

```
#ifdef CONFIG_UART_INTERRUPT_DRIVEN
#define RPI_UART_IRQ_CONFIG_INIT .irq_config_func = ...,
#else
#define RPI_UART_IRQ_CONFIG_INIT
#endif
```

or, make use of `IF_ENABLED(...)`

Reply:
Done!